### ISSUE TITLE
Configuration Validator:  validation of unknown fields

### ISSUE THREAD

**Issue by malex984** [hilbert/hilbert-cli#36]
**@elondaits commented about the validation error due to unknown fields in [hilbert-cli-config.py](https://github.com/hilbert/hilbert-cli/blob/feature/config_validate/config/hilbert-cli-config.py):**
> Also, your cfg could hold settings for things not yet in the code. For instance if you added a feature in v1.3 and then had to rollback to v1.2 but don't want to change the cfg to remove the extra keys added (assuming its backwards-compatible). So validation should be relaxed and check that what is there is OK but simply ignore things it doesn't know about.

**Comment by malex984** [comment]
Ok, then it will not be an error but a warning.

**Comment by malex984** [comment]
**@elondaits reply:**

> Usually not even a warning is needed... I think most systems simply ignore configuration keys they don't "know".

**Comment by malex984** [comment]
**@porst17 comment on https://github.com/hilbert/hilbert-cli/issues/36#issuecomment-259427265:**

> I insist on issuing a warning. If you have a typo in a config key, it will go unnoticed otherwise.

**Comment by malex984** [comment]
What about also adding a CLI option `--quiet` to silence all warnings? 

Note: may be part of general CLI design (#14)

**Comment by elondaits** [comment]
No, either way is OK by me... but doing both with an extra option to choose is the worst of both worlds. We should race to a working program, not create extra problems. If we have extra time we can use it to test and document... and just after that to add extra options.

**Comment by malex984** [comment]
Right now: 
 - in default mode: only warning (`WARNING: Unhandled extra Key: '<keyname>'`) is shown.
 - in `PEDANTIC` mode: error (`ERROR: Unhandled extra Key: '<keyname>'`) is triggered.

TODO: @malex984: add a unit test for this

**Comment by malex984** [comment]
Note: corresponding unit-test will be part of #14

**Comment by porst17** [comment]
To me it looks like `PEDANTIC` mode mixes two things: It turns warnings into errors and does some additional tests (like testing if hosts are actually reachable by ssh). This should be two separate things because you would not be able to run the additional tests if the config validator showed certain warnings before (in non-`PEDANTIC` mode).

Solution: Provide two separate options, one for additional tests and one to turn warnings into errors (like `-Werror` for `gcc`).

**Comment by malex984** [comment]
@porst17 Ok, what about a global switch like  **Skip External Checks**? E.g. `-s / --skipexternal`?

Note: all **internal** checks (which require no execution or external networking) should be fast.

**Comment by porst17** [comment]
Hmmm... something along these lines. I don't like the name that much, though, but I don't have a better alternative. Maybe @elondaits has a good idea.

Also, `-skipexternal` suggests that external tests are performed by default. Is this true?

**Comment by malex984** [comment]
Hm, checks for ssh-aliases are performed **in PEDANTIC mode only** (execute `ssh <address> exit 0`).
Other checks are performed in any case since they can fail quickly:
 - execute some external tool/script or 
 - URI/URL correctness/existence

ps: Later on we can consider switching to some Python library with SSH client (e.g. http://docs.paramiko.org/en/2.1/api/client.html) instead of executing `ssh <address> [options] remote_command` as a subprocess. Also i am not sure whether it is able to deal with `~/.ssh/config` settings and keys on its own...

**Comment by elondaits** [comment]
If hilbert is able to run with a "bad" cfg as long as it doesn't do the pedantic check.... is it really a bad cfg? Why not do whatever it can with it? Why shut down the whole museum if only one station or app is misconfigured? And what if the problem is not in the cfg but in a computer not responding, a repository being corrupted, etc? I don't think checking external resources is equivalent to verifying the cfg.

I dare say that probing the stations to see if they're reachable is overstepping our responsibilities. I don't remember every seeing that in any unix app / server... if they're misconfigured they attempt to run to the best of their abilities and fail if there's a problem.

... We could aim higher than the norm and think that we're doing the right thing by making extra sure if the cfg is OK, but a lot of errors are leaking through. For instance:

- If you configure two stations with the same host name / IP (copy & paste mistake) and they both pass the probe because they're the same.
- If you add the hostname of a test or dev station instead of the right one.
- If you simply forget adding a station, so you never probe it.
- If the test is not thorough enough. The machine is reachable via ssh, but it doesn't have hilbert. Or it's reachable and has hilbert, but hilbert is misconfigured / has bad permissions / has a piece missing / is an older version.

And:

- If it's not reachable it might be simply off. Perhaps it has to be in the cfg but it's broken at the moment, missing a power supply, whatever... The absence of a station should not prohibit fully testing things, nor should it force commenting it out, which means modifying the config which is a sensitive operation since it's the easiest way to add new errors.

- Also admins might want to keep "spare" stations off in a closet but still have them configured for easy reference and so other employees can just "plug and play" without touching the config if they ever need them.

So I think probing the stations doesn't give enough assurances, and it causes some complications. If you want to test the cfg you still have to see if it allows you to control each station... that's the only test that's good enough.

Also, OMD probes the stations and a check to see if hilbert is installed and properly configured could be created there. No?

I'm not against a `hilbert check` command that does ssh to the station and returns a simple sanity check... like a ping of sorts... that could be useful for testing / configuring... but I wouldn't think of it as a final / master validation of the system or config.

**Comment by malex984** [comment]
@porst17 i will have to check the code for the current state...
